src/peft/tuners/adamix.py

```diff
@@ -0,0 +1,416 @@
+            adamix_config.sharing_up,
+            adamix_config.return_two_views,
+            adamix_config.inference_mode,
+            dtype,
+        )
+        return new_module
+
+    def _check_target_module_exists(self, adamix_config, key):
+        if isinstance(adamix_config.target_modules, str):
+            target_module_found = True if adamix_config.target_modules == key else False
+        else:
+            for target in adamix_config.target_modules:
+                target_module_found = True if target == key else False
+        return target_module_found
+
+    def get_param_dtype(self, target):
+        params = list(target.parameters())
+        pos = 0
+        while pos < len(params):
+            if params[pos] is not None:
+                dtype = params[pos].dtype
+                return dtype
+        raise ValueError("All parameters in the model are None")
+
+    def _find_and_replace(self, adapter_name):
+        adamix_config = self.peft_config[adapter_name]
+        self._check_quantization_dependency()
+        is_target_modules_in_base_model = False
+
+        module_dict = dict(self.model.named_modules())
+        for key in module_dict:
+            module_dict[key] = type(module_dict[key]).__name__
+            if not self._check_target_module_exists(adamix_config, module_dict[key]):
+                continue
+
+            if not is_target_modules_in_base_model:
+                is_target_modules_in_base_model = True
+            _, target, _ = _get_submodules(self.model, key)
+
+            dtype = self.get_param_dtype(target)
+            new_module = self._create_new_module(adamix_config, adapter_name, target, dtype)
+            self._add_adapter_module(target, new_module)
+
+        if not is_target_modules_in_base_model:
+            raise ValueError(
+                f"Target modules {adamix_config.target_modules} not found in the base model. "
+                f"Please check the target modules and try again."
+            )
+
+    @staticmethod
+    def _is_valid_match(key: str, target_key: str):
+        """
+        Helper function to match module names target_key and key. Makes sure that either the key is exactly the
+        target_key or the target_key is a submodule of key
+        """
+        if key.endswith(target_key):
+            if len(key) > len(target_key):
+                return key.endswith("." + target_key)  # must be a sub module
+            return True
+        return False
+
+    @staticmethod
+    def composite_forward(f_out, f_in):
+        def wrapper(hidden_states, *args, **kwargs):
+            tuple_input = False
+            hidden_states = f_in(hidden_states, *args, **kwargs)
+            if isinstance(hidden_states, tuple):
+                tuple_input = True
+                hidden_states, other_args = hidden_states[0], hidden_states[1:]
+            hidden_states = f_out(hidden_states)
+            if tuple_input:
+                return (hidden_states,) + other_args
```

Comment:
This could break downstream code if the type is a subtype of tuple and not exactly tuple. E.g. if the type is a `namedtuple`, `isinstance(hidden_states, tuple` would be `True` but the output would be a different type.

Reply:
That is one of the problems with the AdaMix model. Unlike Lora, where you can replace any layer with new low rank layers, adamix can only add the adapter after the selected layer. As a result, without knowing the exact class/function definition of the model, AdaMix is restricted. But I think this can be addressed in another PR